custom/sensors/TalonEncoder.java

```diff
@@ -0,0 +1,89 @@
+package org.usfirst.frc4904.standard.custom.sensors;
+
+
+import edu.wpi.first.wpilibj.CANTalon;
+import edu.wpi.first.wpilibj.PIDSourceType;
+
+public class TalonEncoder implements CustomEncoder {
+	protected final CANTalon talon;
+	private PIDSourceType pidSource;
+	private double distancePerPulse;
+	private boolean reverseDirection;
+	private boolean rateMode;
+	
+	public TalonEncoder(String name, CANTalon talon, boolean reverseDirection, double distancePerPulse, boolean rateMode) {
+		this.talon = talon;
+		this.reverseDirection = reverseDirection;
+		this.distancePerPulse = distancePerPulse;
+		this.rateMode = rateMode;
+		setPIDSourceType(PIDSourceType.kDisplacement);
+	}
+	
+	@Override
+	public void setPIDSourceType(PIDSourceType pidSource) {
+		this.pidSource = pidSource;
+	}
+	
+	@Override
+	public PIDSourceType getPIDSourceType() {
+		return pidSource;
+	}
+	
+	@Override
+	public double pidGet() {
+		if (rateMode) {
+			return getRate();
+		}
+		return getDistance();
+	}
+	
+	@Override
+	public int get() {
+		return talon.getEncPosition();
+	}
+	
+	@Override
+	public double getDistance() {
+		if (reverseDirection) {
+			return distancePerPulse * (double) talon.getEncPosition() * -1.0;
+		} else {
+			return distancePerPulse * (double) talon.getEncPosition();
+		}
+	}
+	
+	@Override
+	public boolean getDirection() {
+		return !reverseDirection == (talon.getEncVelocity() >= 0);
+	}
+	
+	@Override
+	public boolean getStopped() {
+		return Math.abs(getRate()) <= 0.0001;
```

Comment:
In the interest of limiting magic numbers, can we make this a constant?
Also, this needs to be way smaller and hopefully less arbitrary. Would Double.MIN_VALUE, Double.MIN_NORMAL, or Double.MIN_EXPONENT work?
Maybe we should have a system-wide epsilon.
Finally, we need to work on eliminating floating point errors from the other end. See Mark Jeronimus's answer [here](http://stackoverflow.com/questions/8194917/comparing-a-double-against-zero) - the jist is, do all of your calculations for the numerator and denominator and divide only once. It might make the calculations a little (read: a lot) uglier, though.